ykrt/src/compile/jitc_yk/codegen/x64/mod.rs

```diff
@@ -4647,23 +4719,34 @@ mod tests {
               entry:
                 %0: i8 = param reg
                 %1: i8 = param reg
-                %2: i1 = eq %0, 2i8
-                %3: i8 = and %1, 3i8
-                %4: i1 = eq %3, 4i8
-                black_box %2
-                black_box %4
+                %2: i64 = param reg
+                %3: i1 = eq %0, 2i8
+                %4: i8 = and %1, 3i8
+                %5: i1 = eq %4, 4i8
+                %6: i64 = and %2, 2147483647i64
+                %7: i64 = and %2, 2147483648i64
+                black_box %3
+                black_box %5
+                black_box %6
+                black_box %7
             ",
             "
                 ...
-                ; %2: i1 = eq %0, 2i8
+                ; %3: i1 = eq %0, 2i8
                 and r.32.a, 0xff
                 cmp r.32.a, 0x02
                 setz ...
-                ; %3: i8 = and %1, 3i8
+                ; %4: i8 = and %1, 3i8
                 and r.32.b, 0x03
-                ; %4: i1 = eq %3, 4i8
+                ; %5: i1 = eq %4, 4i8
                 cmp r.32.b, 0x04
                 setz ...
+                ; %6: i64 = and %2, 2147483647i64
+                mov r.64.x, r.64.y
+                and r.64.y, 0x7fffffff
+                ; %7: i64 = and %2, 2147483648i64
```

Comment:
I think this is what I wanted to see. The moment the top bit is set, we no longer use `and r64, imm32` because the sign extend would give incorrect semantics.

Looks like you checked one value either side of the threshold. Did I understand correctly?

Reply:
> Did I understand correctly?

Yep.